itest/rust/src/rid_test.rs

```diff
@@ -0,0 +1,129 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{collections::HashSet, thread};
+
+use godot::{
+    engine::{Engine, RenderingServer},
+    prelude::{inner::InnerRid, Color, Rid, Vector2},
+};
+
+use crate::itest;
+
+/// Disable printing errors from Godot. Ideally we should catch and handle errors, ensuring they happen.
+/// But that isn't possible, so for now we just disable printing the error to avoid spamming the terminal.
+fn should_error(mut f: impl FnMut()) {
+    Engine::singleton().set_print_error_messages(false);
+    f();
+    Engine::singleton().set_print_error_messages(true);
+}
+
+#[itest]
+fn rid_equiv() {
+    let invalid: Rid = Rid::Invalid;
+    let valid: Rid = Rid::new((10 << 32) | 20);
+    assert!(!InnerRid::from_outer(&invalid).is_valid());
+    assert!(InnerRid::from_outer(&valid).is_valid());
+
+    assert_eq!(InnerRid::from_outer(&invalid).get_id(), 0);
+    assert_eq!(InnerRid::from_outer(&valid).get_id(), (10 << 32) | 20);
+}
+
+#[itest]
+fn canvas_set_parent() {
+    // This originally caused UB, but still testing it here in case it breaks.
+    let mut server = RenderingServer::singleton();
+    let canvas = server.canvas_create();
+    let viewport = server.viewport_create();
+
+    should_error(|| server.canvas_item_set_parent(viewport, canvas));
+    should_error(|| server.canvas_item_set_parent(viewport, viewport));
+
+    server.free_rid(canvas);
+    server.free_rid(viewport);
+}
+
+#[itest]
+fn multi_thread_test() {
+    let threads = (0..10)
+        .map(|_| {
+            thread::spawn(|| {
+                let mut server = RenderingServer::singleton();
+                let mut v = vec![];
+                for _ in 0..1000 {
+                    v.push(server.canvas_item_create());
+                }
+                v
```

Comment:
This could probably be simplfied by [mapping a range](https://stackoverflow.com/a/48021563) 🙂